cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll

```diff
@@ -602,9 +602,14 @@ private predicate exprToExprStep_nocfg(Expr fromExpr, Expr toExpr) {
       exists(DataFlowFunction f, FunctionInput inModel, FunctionOutput outModel, int iIn |
         call.getTarget() = f and
         f.hasDataFlow(inModel, outModel) and
-        outModel.isReturnValue() and
-        inModel.isParameter(iIn) and
-        fromExpr = call.getArgument(iIn)
+        fromExpr = call.getArgument(iIn) and
+        (
+          inModel.isParameter(iIn) and
+          outModel.isReturnValue()
+          or
+          inModel.isParameterDeref(iIn) and
+          outModel.isReturnValueDeref()
```

Comment:
I'd argue that there should not be data flow from `x` to `strdup(x)`, only taint flow. The distinction between data and taint flow remains a bit blurry, and in the end it's about what people might use the libraries for. I can imagine using data flow to follow a pointer from `malloc` to `free`. Then I'd not want to follow it across `strdup`.

Reply:
I've moved it to taint tracking - and then realized that it's the same as one of the cases in `exprToDefinitionByReferenceStep` only to a ReturnValueDeref rather than a ParameterDeref ... and in fact all of the cases in `exprToDefinitionByReferenceStep` should be duplicated for return values in the same way.

Reply:
That seems right to me, but I'd like to hear the opinion of @rdmarsh2 before merging.

@rdmarsh2, I think you were involved the last time we made the taint-tracking library use the "deref" data-flow models to conflate pointers with their objects. We did it only for reference arguments, and this PR now extends it it to cover return values as well.

Reply:
There's the same caveats about false positives, and I'd like to see some testing on real code, but I'm for it.

Reply:
Seems right to me as well. It does mean we should test this PR more carefully.